### PR TITLE
Try fix incorrect "explicit lifetime name needed"

### LINKED ISSUE #65285: Incorrect "explicit lifetime name needed" error
```rust
struct Nested<K>(K);

trait X<'a, K: 'a> {
    fn foo<'b, L: X<&'b Nested<K>>>();
}
```
produces:
```
error[E0637]: `&` without an explicit lifetime name cannot be used here
 --> src/lib.rs:4:19
  |
4 |     fn foo<'b, L: X<&'b Nested<K>>>();
  |                   ^^^^^^^^^^^^^^^^ explicit lifetime name needed here
```
despite there being no unnamed `&`.

### PR DESCRIPTION
**Pull request by Phosphorus15**
This pr is trying to fixes #65285 .

### CODE REVIEW

**Comment by varkor** [comment]
Can you add a test with the new diagnostic?

**Comment by Phosphorus15** [comment]
Sure yes, there ought to be a test correspond to this issue.

**Comment by Phosphorus15** [comment]
It seems enough to solve this problem, please review 😃

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -3260,6 +3260,17 @@ impl<'a> LoweringContext<'a> {
         self.new_named_lifetime(id, span, hir::LifetimeName::Error)
     }
 
+    /// Report an error on illegal use of `'_` within paths
+    /// Or elide the lifetime parameter (if possible) for the lifetime resolver to check
+    /// Refer to issue #65285
+    /// return a "error lifetime".
+    fn new_missing_lifetime(&mut self, id: Option<NodeId>, span: Span) -> hir::Lifetime {
+        match id {
```
Comment (varkor) [inline src/librustc/hir/lowering.rs]: This could be `if id.is_some() { … }`.

File: src/librustc/hir/lowering.rs [hunk 2]
```diff
@@ -3284,7 +3295,7 @@ impl<'a> LoweringContext<'a> {
             // This is the normal case.
             AnonymousLifetimeMode::PassThrough => self.new_implicit_lifetime(span),
 
-            AnonymousLifetimeMode::ReportError => self.new_error_lifetime(None, span),
+            AnonymousLifetimeMode::ReportError => self.new_missing_lifetime(None, span),
```
Comment (varkor) [inline src/librustc/hir/lowering.rs]: This is always called with `None`, so you might as well just use `new_implicit_lifetime` here, no?

Comment (Phosphorus15) [reply src/librustc/hir/lowering.rs]: Well, shame of me. Thanks for pointing out XD

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -3283,8 +3283,10 @@ impl<'a> LoweringContext<'a> {
             }
             // This is the normal case.
             AnonymousLifetimeMode::PassThrough => self.new_implicit_lifetime(span),
-
-            AnonymousLifetimeMode::ReportError => self.new_error_lifetime(None, span),
+            // elide the lifetime parameter (if possible) for the lifetime resolver to check
```
Comment (varkor) [inline src/librustc/hir/lowering.rs]: It's not clear to me from this comment why this change works. Could you try to elaborate what this does in your comment?

File: src/librustc/hir/lowering.rs [hunk 4]
```diff
@@ -3281,10 +3281,13 @@ impl<'a> LoweringContext<'a> {
                 let id = self.sess.next_node_id();
                 self.new_named_lifetime(id, span, hir::LifetimeName::Error)
             }
-            // This is the normal case.
-            AnonymousLifetimeMode::PassThrough => self.new_implicit_lifetime(span),
-
-            AnonymousLifetimeMode::ReportError => self.new_error_lifetime(None, span),
+            // `PassThrough` is the normal case.
+            // We merged `ReportError` with this branch (Refer to issue #65285)
```
Comment (varkor) [inline src/librustc/hir/lowering.rs]:
```suggestion
            // `new_error_lifetime`, which would usually be used in the case of `ReportError`,
            // is unsuitable here, as these can occur from missing lifetime parameters in a
            // `PathSegment`, for which there is no associated `'_` or `&T` with no explicit
            // lifetime. Instead, we simply create an implicit lifetime, which will be checked
            // later, at which point a suitable error will be emitted.
```
Just some tweaks to the wording. I think this is clearer.

**Comment by varkor** [comment]
r=me with comment suggestion and squashing the commits.

**Review by varkor — COMMENTED** [review]
Why remove the full stop here?

**Comment by Phosphorus15** [comment]
> 
> 
> Why remove the full stop here?

There were something wrong with CI, and I just tried to trigger the re-run by a commit, I'll revert this change

**Comment by varkor** [comment]
In the future, you can trigger the CI to restart by amending the last commit and force-pushing.

If you could squash your commits into a single commit, I'll approve the PR :)

**Comment by Phosphorus15** [comment]
Thanks a lot 🎁 , I'm somewhat new to the rust community and you've helped me greatly.

**Pull request merged**